src/hardware/mixer.cpp

```diff
@@ -1693,46 +1694,140 @@ AudioFrame MixerChannel::ApplyCrossfeed(const AudioFrame frame) const
+	};
+	// Disable fadeout (default)
+	if (has_false(prefs)) {
+		wants_fadeout = false;
+		return;
+	}
+	// Enable fadeout with defaults
+	if (has_true(prefs)) {
+		set_wait_and_fade(default_wait_ms, default_wait_ms);
+		wants_fadeout = true;
 		return;
 	}
 
-	constexpr auto silence_threshold = 1.0f;
+	// Let the fade out last between 10 ms and 3 seconds.
+	constexpr auto min_fade_ms = 10;
+	constexpr auto max_fade_ms = 3000;
 
-	had_noise = fabsf(frame.left) > silence_threshold ||
-	            fabsf(frame.right) > silence_threshold;
+	// Custom setting in 'WAIT FADE' syntax, where both are milliseconds.
+	if (auto prefs_vec = split(prefs); prefs_vec.size() == 2) {
+		const auto wait_ms = parse_int(prefs_vec[0]);
+		const auto fade_ms = parse_int(prefs_vec[1]);
+		if (wait_ms && fade_ms) {
+			const auto wait_is_valid = (*wait_ms >= min_wait_ms &&
+			                            *wait_ms <= max_wait_ms);
+
+			const auto fade_is_valid = (*fade_ms >= min_fade_ms &&
+			                            *fade_ms <= max_fade_ms);
+
+			if (wait_is_valid && fade_is_valid) {
+				set_wait_and_fade(*wait_ms, *fade_ms);
+				wants_fadeout = true;
+				return;
+			}
+		}
+	}
+	// Otherwise inform the user and disable the fade
+	LOG_WARNING("%s: Invalid custom fadeout definition: '%s'. Must be "
+	            "specified in \"WAIT FADE\" format where WAIT is between "
+	            "%d and %d (in milliseconds) and FADE is between %d and "
+	            "%d (in milliseconds). Using 'off'",
+	            channel.GetName().c_str(),
+	            prefs.data(),
+	            min_wait_ms,
+	            max_wait_ms,
+	            min_fade_ms,
+	            max_fade_ms);
+
+	wants_fadeout = false;
 }
 
-void MixerChannel::Sleeper::MaybeSleep()
+void MixerChannel::Sleeper::DecrementFadeLevel(const int64_t awake_for_ms)
 {
-	constexpr auto consider_sleeping_after_ms = 250;
+	assert(awake_for_ms >= fadeout_or_sleep_after_ms);
+	const auto elapsed_fade_ms = static_cast<float>(
+	        awake_for_ms - fadeout_or_sleep_after_ms);
 
-	// Not enough time has passed.. try again later
-	if (GetTicksSince(woken_at_ms) < consider_sleeping_after_ms) {
-		return;
+	const auto decrement = fadeout_decrement_per_ms * elapsed_fade_ms;
+
+	constexpr auto min_level = 0.0f;
+	constexpr auto max_level = 1.0f;
+	fadeout_level = std::clamp(max_level - decrement, min_level, max_level);
+}
+
+MixerChannel::Sleeper::Sleeper(MixerChannel& c, const uint16_t sleep_after_ms)
+        : channel(c),
+          fadeout_or_sleep_after_ms(sleep_after_ms)
+{
+	// The constructed sleep period is programmatically controlled (so assert)
+	assert(fadeout_or_sleep_after_ms >= min_wait_ms);
+	assert(fadeout_or_sleep_after_ms <= max_wait_ms);
+}
+
+// Either fades the frame by the current fade-level, or records if the frame had
+// a magnitude greater than 1. Once "had_signal" is tripped, it can only be
+// reset after waking up.
+void MixerChannel::Sleeper::MaybeFadeOrListen(AudioFrame& frame)
```

Comment:
You don't need to change this specific instance if you don't wanna, but instead of `foo(Type& a)` we can do `a foo(const Type& a)` and return the new value instead of mutating it. Because of the mandatory optimisations of new C++ compilers that has the same performance and the calculation would be performed in place (I'm pretty sure of that), but it's a nicer signature and then it's a pure function.

Reply:
Switch to a return value.
In the case where we're just sleeping, the frame is simply passed through.